burn-core/src/optim/rmsprop.rs

```diff
@@ -0,0 +1,528 @@
+use crate::{
+    self as burn, grad_clipping::GradientClippingConfig, module::ADModule, record::Record,
+    LearningRate,
+};
+
+use super::{
+    decay::{WeightDecay, WeightDecayConfig, WeightDecayState},
+    SimpleOptimizer,
+};
+use crate::config::Config;
+use crate::optim::adaptor::OptimizerAdaptor;
+use crate::tensor::{backend::ADBackend, Tensor};
+use burn_tensor::backend::Backend;
+
+/// Configuration to create the [RMSProp](RMSProp) optimizer.
+#[derive(Config)]
+pub struct RMSPropConfig {
+    /// Smoothing constant.
+    #[config(default = 0.99)]
+    alpha: f32,
+    /// momentum for RMSProp.
+    #[config(default = 0.9)]
+    momentum: f32,
+    /// A value required for numerical stability.
+    #[config(default = 1e-5)]
+    epsilon: f32,
+    /// if True, compute the centered RMSProp, the gradient is normalized by an estimation of its variance
+    #[config(default = false)]
+    centered: bool,
+    /// [Weight decay](WeightDecayConfig) config.
+    weight_decay: Option<WeightDecayConfig>,
+    /// [Gradient Clipping](GradientClippingConfig) config.
+    grad_clipping: Option<GradientClippingConfig>,
+}
+
+impl RMSPropConfig {
+    /// Initialize RMSProp optimizer.
+    ///
+    /// # Returns
+    ///
+    /// Returns an optimizer that can be used to optimize a module.
+    pub fn init<B: ADBackend, M: ADModule<B>>(
+        &self,
+    ) -> OptimizerAdaptor<RMSProp<B::InnerBackend>, M, B> {
+        let weight_decay = self.weight_decay.as_ref().map(WeightDecay::new);
+
+        let mut optim = OptimizerAdaptor::from(RMSProp {
+            alpha: self.alpha,
+            centered: self.centered,
+            weight_decay: weight_decay,
+            momentum: RMSPropMomentum {
+                momentum: self.momentum,
+                epsilon: self.epsilon,
+            },
+        });
+
+        if let Some(config) = &self.grad_clipping {
+            optim = optim.with_grad_clipping(config.init());
+        }
+
+        optim
+    }
+}
+
+/// Optimizer that implements stochastic gradient descent with momentum.
+/// The optimizer can be configured with [RMSPropConfig](RMSPropConfig).
+pub struct RMSProp<B: Backend> {
+    alpha: f32,
+    // epsilon: f32,
+    centered: bool,
+    // momentum: Option<Momentum<B>>,
+    momentum: RMSPropMomentum,
+    weight_decay: Option<WeightDecay<B>>,
+}
+
+/// State of [RMSProp](RMSProp)
+#[derive(Record, Clone, new)]
+pub struct RMSPropState<B: Backend, const D: usize> {
+    weight_decay: Option<WeightDecayState<B, D>>,
+    square_avg: Option<SquareAvgState<B, D>>,
+    centered: Option<CenteredState<B, D>>,
+    momentum: Option<RMSPropMomentumState<B, D>>,
```

Comment:
Is it necessary to have those states wrapped into options?

Reply:
Oh, I didn't realize that I can shadow these intermediate variables in the `step` method before. I changed it now. And also add a match sentence to avoid always assigning zero tensors to `grad_avg` every time it calls the centered transform with `centered` is false.